File: backend/app/extractors/mfg_date.py
"""Extract manufacture / packing date per Rule 6(1)(d)."""

from __future__ import annotations

import re

from app.domain import ExtractedField, ImageMeta, OCRWord
from app.extractors.base import merge_bboxes

DATE_PATTERN = re.compile(
    r"(?i)\b(?:mfg|mfd|manufactured|packed|pkd|pkd\.?|mfg\.?)"
    r"[:\s,.]*"
    r"("
    r"(?:0?[1-9]|1[0-2])[\/\-\s]\d{2,4}"
    r"|"
    r"(?:jan|feb|mar|apr|may|jun|jul|aug|sep|sept|oct|nov|dec)[a-z]*\s+\d{4}"
    r")"
)


def extract_mfg_date(
    ocr_words: list[OCRWord],
    image_meta: ImageMeta,  # noqa: ARG001
    date_regex: str,
) -> ExtractedField | None:
    """Find manufacture dates in a single OCR token or adjacent OCR words."""
    date_re = re.compile(date_regex, re.IGNORECASE)
    text = " ".join(word.text for word in ocr_words)
    match = date_re.search(text)
    if match is None:
        return ExtractedField(
            name="mfg_date", value=None, bbox=None, confidence=0.0, evidence_spans=[]
        )

    matched_words = [
        word
        for word in ocr_words
        if match.start() <= text.find(word.text) + len(word.text)
        and text.find(word.text) <= match.end()
    ]
    if not matched_words:
        return ExtractedField(
            name="mfg_date", value=None, bbox=None, confidence=0.0, evidence_spans=[]
        )

    return ExtractedField(
        name="mfg_date",
        value=match.group(1).strip() if match.lastindex else match.group(0).strip(),
        bbox=merge_bboxes(matched_words),
        confidence=sum(word.confidence for word in matched_words) / len(matched_words),
        evidence_spans=[word.bbox for word in matched_words],
    )


__all__ = ["extract_mfg_date", "DATE_PATTERN"]

File: frontend/lib/rules/extractors/mfg-date.ts
import type { ExtractedField, ImageMeta, OCRWord } from '../domain';
import { avgConfidence, compileRegex, mergeBboxes } from './base';
import { groupWordsIntoSpatialRows } from '../../spatial-layout';

// Standard Indian date pattern: DD/MM/YYYY, DD-MM-YYYY, DD.MM.YYYY, MM/YYYY, DD Mon YYYY, Mon YYYY
const BROAD_DATE_PATTERN =
  /\b((?:(?:0?[1-9]|[12][0-9]|3[01])[\/\-\.](?:0?[1-9]|1[0-2])[\/\-\.]\d{2,4})|(?:(?:0?[1-9]|1[0-2])[\/\-\.]\d{2,4})|(?:(?:0?[1-9]|[12][0-9]|3[01])[\s\/\-\.]*(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*[\s\/\-\.]*\d{2,4})|(?:(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*[\s\/\-\.]*\d{2,4}))\b/i;

const MFG_PREFIX =
  /\b(?:date\s+of\s+manufacture|date\s+of\s+mfg|mfg\.?\s*date|mfd\.?\s*date|manufactured\s+on|manufacture\s+date|packaging\s+date|packed\s+on|pkd\.?\s*date|mfg|mfd|packed|pkd)\b/i;

export function extractMfgDate(
  ocrWords: OCRWord[],
  imageMeta: ImageMeta,
  dateRegexStr: string
): ExtractedField {
  const dateRe = compileRegex(dateRegexStr, 'i');
  const fullText = ocrWords.map((w) => w.text).join(' ');

  // Pass 1: Standard rule regex on full text
  const match = dateRe.exec(fullText);
  if (match) {
    const wordSpans: Array<{ word: OCRWord; start: number; end: number }> = [];
    let currentPos = 0;
    for (const word of ocrWords) {
      const start = currentPos;
      const end = start + word.text.length;
      wordSpans.push({ word, start, end });
      currentPos = end + 1;
    }

    const matchStart = match.index;
    const matchEnd = match.index + match[0].length;
    const matchedWords = wordSpans
      .filter(({ start, end }) => !(end < matchStart || start > matchEnd))
      .map(({ word }) => word);

    if (matchedWords.length > 0) {
      const val = (match[1] ?? match[0]).trim();
      const conf =
        matchedWords.reduce((sum, w) => sum + w.confidence, 0) / matchedWords.length;
      console.log(`[extractMfgDate] Matched via fullText regex: "${val}"`);
      return {
        name: 'mfg_date',
        value: val,
        bbox: mergeBboxes(matchedWords),
        confidence: conf,
        evidence_spans: matchedWords.map((w) => w.bbox),
      };
    }
  }

  // Pass 2: Spatial Row Search (handles two-column "DATE OF MANUFACTURE" [Col 1] -> "13/05/2026" [Col 2])
  const spatialRows = groupWordsIntoSpatialRows(ocrWords);
  for (const row of spatialRows) {
    if (MFG_PREFIX.test(row.text)) {
      console.log(`[extractMfgDate] Checking spatial row: "${row.text}"`);
      const dateMatch = BROAD_DATE_PATTERN.exec(row.text);
      if (dateMatch) {
        const val = dateMatch[1].trim();
        console.log(`[extractMfgDate] Matched spatial row date: "${val}" in "${row.text}"`);
        return {
          name: 'mfg_date',
          value: val,
          bbox: row.bbox,
          confidence: avgConfidence(row.words),
          evidence_spans: row.words.map((w) => w.bbox),
        };
      }
    }
  }

  console.log('[extractMfgDate] No Mfg Date found in evidence.');
  return {
    name: 'mfg_date',
    value: null,
    bbox: null,
    confidence: 0.0,
    evidence_spans: [],
  };
}
